Add support for new modes in Tuya Climate

moriahmorgan opened the pull request:
# What does this implement/fix?
Added support for Fan Only Mode, Dry Mode, Swing Mode and Fan Speed Control.

Also added/fixed support for entity states syncing with current operation mode.

<!-- Quick description and explanation of changes -->

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>
https://github.com/esphome/feature-requests/issues/2329
**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3096

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
climate:
  - platform: tuya
    id: "tuya_climate"
    name: None
    switch_datapoint: 1
    target_temperature_datapoint: 2
    current_temperature_datapoint: 3
    supports_heat: True
    supports_cool: True
    active_state:
      datapoint: 4
      cooling_value: 0
      heating_value: 1
      drying_value: 2
      fanonly_value: 3
    preset:
      eco:
        datapoint: 8
        temperature: 28
      sleep:
        datapoint: 101
    swing_mode:
      vertical_datapoint: 106
      horizontal_datapoint: 107
    fan_mode:
      datapoint: 5
      auto_value: 0
      low_value: 2
      medium_value: 3
      middle_value: 4
      high_value: 1

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [?] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in esphome/esphome-docs#3096

moriahmorgan committed:
Add support support for new modes

Added support for Fan Only Mode, Dry Mode, Swing Mode and Fan Speed Control.

Also added/fixed support for entity states syncing with current operation mode.

moriahmorgan committed:
Add support for more climate modes in climate.tuya

Added support for Fan Only Mode, Dry Mode, Swing Mode and Fan Speed Control.

Also added/fixed support for entity states syncing with current operation mode.

This commit fixes the namespace, because I uploaded the test files to start with.

moriahmorgan committed:
Code Formatting Changes per Clang format.

moriahmorgan committed:
More clang formatting fixes.

moriahmorgan committed:
Breaking Change: Group YAML entries by type

Add grouping to Preset, Swing Mode, Fan Speed and Active State. This is a breaking change.

blacktirion commented:
This is now a breaking change. Added grouping to Preset, Swing Mode, Fan Speed and Active State.

moriahmorgan committed:
Formatting Changes for validation

Formatting changes to be compliant with black and flake8. Also changed constants to match expected format.

moriahmorgan committed:
More constant value fixes

moriahmorgan committed:
Final black formatting check?

jesserockz commented on esphome/components/tuya/climate/__init__.py:
```diff
@@ -99,27 +153,25 @@ def validate_eco_values(value):
-    cv.has_at_most_one_key(CONF_ACTIVE_STATE_DATAPOINT, CONF_COOLING_STATE_PIN),
-    validate_eco_values,
+    validate_cooling_values,
+    cv.has_at_most_one_key(CONF_STATE_DATAPOINT, CONF_HEATING_STATE_PIN),
+    cv.has_at_most_one_key(CONF_STATE_DATAPOINT, CONF_COOLING_STATE_PIN),
```
```suggestion
    cv.has_at_most_one_key(CONF_ACTIVE_STATE, CONF_HEATING_STATE_PIN),
    cv.has_at_most_one_key(CONF_ACTIVE_STATE, CONF_COOLING_STATE_PIN),
```

moriahmorgan replied:
Is it OK to commit all these at one time with a new file version that has all the changes requested here? I've also resolved some issues I saw with certain code validation with COOLING_STATE_PIN and ACTIVE_STATE COOLING_VALUE, that the function wasn't working as expected. And I moved Sleep Datapoint under its own block in presets.

moriahmorgan replied:
Resolved in latest commit.

jesserockz commented on esphome/components/tuya/climate/__init__.py:
```diff
@@ -99,27 +153,25 @@ def validate_eco_values(value):
-            cv.Optional(CONF_ACTIVE_STATE_DATAPOINT): cv.uint8_t,
-            cv.Optional(CONF_ACTIVE_STATE_HEATING_VALUE, default=1): cv.uint8_t,
-            cv.Optional(CONF_ACTIVE_STATE_COOLING_VALUE): cv.uint8_t,
+            cv.Optional(CONF_ACTIVE_STATE): ACTIVE_STATES,
```
Don't apply this suggestion as it requires more changes, but for removing config options, we switch them to `cv.invalid` do people get nicer errors when they try to compile and remove them in a later version.

```suggestion
            cv.Optional(CONF_ACTIVE_STATE): ACTIVE_STATES,
            cv.Optional("active_state_datapoint"): cv.invalid("'active_state_datapoint' has moved inside of the 'active_state' config block as 'datapoint'"),
            cv.Optional("active_state_heating_value"): cv.invalid(...),
            cv.Optional("active_state_cooling_value"): cv.invalid(...),
```

jesserockz commented on esphome/components/tuya/climate/__init__.py:
```diff
@@ -67,30 +83,68 @@ def validate_temperature_multipliers(value):
-    return value
+ACTIVE_STATES = cv.Schema(
+    {
+        cv.Required(CONF_STATE_DATAPOINT): cv.uint8_t,
```
Lets simplify these as just `datapoint` in config since they are already inside the block that describes what it is for.

Same for eco and fan

```suggestion
        cv.Required(CONF_DATAPOINT): cv.uint8_t,
```

jesserockz commented on esphome/components/tuya/climate/__init__.py:
```diff
@@ -67,30 +83,68 @@ def validate_temperature_multipliers(value):
 
+PRESETS = cv.Schema(
+    {
+        cv.Optional(CONF_ECO): {
+            cv.Required(CONF_ECO_DATAPOINT): cv.uint8_t,
+            cv.Optional(CONF_ECO_TEMPERATURE): cv.temperature,
+        },
```
Same reason as above
```suggestion
        cv.Optional(CONF_ECO): {
            cv.Required(CONF_DATAPOINT): cv.uint8_t,
            cv.Optional(CONF_TEMPERATURE): cv.temperature,
        },
```

moriahmorgan committed:
Merge branch 'dev' into tuya_climate_updates

moriahmorgan committed:
Changes to init.py according to reviewer requests

Make changes to _init_.py according to https://github.com/esphome/esphome/pull/5159/files/649b923804c05fa6ee750e824e3d7d70fadeabd9#r1278620976, https://github.com/esphome/esphome/pull/5159/files/649b923804c05fa6ee750e824e3d7d70fadeabd9#r1278621039, https://github.com/esphome/esphome/pull/5159/files/649b923804c05fa6ee750e824e3d7d70fadeabd9#r1278620904, and https://github.com/esphome/esphome/pull/5159/files/649b923804c05fa6ee750e824e3d7d70fadeabd9#r1278620549

Also put Sleep preset in its own config block to be consistent with other presets and fix logic for validate_cooling_values function to better align with existing documentation.

blacktirion commented:
Requested changes reviewed and changes applied in single commit so that I could validate testing. Also put Sleep preset in its own config block to be consistent with other presets and fix logic for validate_cooling_values function to better align with existing documentation.

jesserockz commented on esphome/components/tuya/climate/__init__.py:
```diff
@@ -99,27 +157,40 @@ def validate_eco_values(value):
+            cv.Optional("active_state_datapoint"): cv.invalid(
+                "'active_state_datapoint' has been moved inside of the 'active_state' config block as 'datapoint'"
+            ),
+            cv.Optional("heating_state_datapoint"): cv.invalid(
```
This is not a datapoint (also for cooling, and the message itself
```suggestion
            cv.Optional("active_state_heating_value"): cv.invalid(
```

moriahmorgan replied:
Thanks for the oversight. Missed that one.

moriahmorgan committed:
Commit reviewed change

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

moriahmorgan committed:
update deprecated config option wording

jesserockz commented on esphome/components/tuya/climate/tuya_climate.cpp:
```diff
@@ -129,6 +178,106 @@ void TuyaClimate::control(const climate::ClimateCall &call) {
+    switch (swing_mode) {
+      case climate::CLIMATE_SWING_OFF:
+        if (swing_vertical_ || swing_horizontal_) {
+          swing_vertical_ = false;
+          swing_horizontal_ = false;
```
Member variables should always be prefixed with `this->` please.
```suggestion
          this->swing_vertical_ = false;
          this->swing_horizontal_ = false;
```

moriahmorgan replied:
changed. Thank you for this. Total oversight on my part.

moriahmorgan committed:
add "this->" to member variables that were missed

adding "this->" to some member variables in the swing_mode function.

moriahmorgan committed:
Update _init_.py to use Python 3.8 Walrus operator

Adding Walrus Operator in the to_code function for _init_.py similar to https://github.com/esphome/esphome/pull/5181

blacktirion commented:
Not sure what's causing `CI / Run YAML test 4` (pull_request) to fail...

moriahmorgan committed:
Merge branch 'dev' into tuya_climate_updates

moriahmorgan committed:
Fix Temperature_Multiplier config entry for code generation

shakin89 commented:
Is there any chance to have this pr merged?
I successfully flashed 5 devices, Beca BAC-002 CB3S based, and cannot set correctly fan speed during heating/cooling nor set fan_only mode. 
I would be sad to come back to the original firmware.

moriahmorgan commented:
> Is there any chance to have this pr merged? I successfully flashed 5 devices, Beca BAC-002 CB3S based, and cannot set correctly fan speed during heating/cooling nor set fan_only mode. I would be sad to come back to the original firmware.

I think it's just waiting on me to finish documentation. I have had a bit of a busy last few months health related and it keeps getting put on the back burner. I will attempt to finish this weekend, since @jesserockz is currently out for the rest of the week.

shakin89 commented:
> I think it's just waiting on me to finish documentation. I have had a bit of a busy last few months health related and it keeps getting put on the back burner. I will attempt to finish this weekend, since @jesserockz is currently out for the rest of the week.

I'm sorry, hope you are well now. Let me know if i can help writing the docs in some way. 
Thanks both for the great work done till now.

moriahmorgan commented:
@shakin89 Docs updated, just waiting for review, approval, then merge by @jesserockz when he has time, unless there are other issues to resolve. It has been some months since I looked at the code and I don't know if there's any issues right now with merging this.

kbx81 committed:
Merge branch 'dev' into tuya_climate_updates
